Framework/Algorithms/src/CalculatePolynomialBackground.cpp

```diff
@@ -0,0 +1,279 @@
+                                     const Mantid::API::MatrixWorkspace &ws,
+                                     const size_t wsIndex) {
+  const auto minX = ws.x(wsIndex).front();
+  const auto maxX = ws.x(wsIndex).back();
+  if (ranges.empty()) {
+    return std::pair<double, double>(minX, maxX);
+  }
+  const auto minmaxIt = std::minmax_element(ranges.cbegin(), ranges.cend());
+  const auto minEdge = *minmaxIt.first;
+  const auto maxEdge = *minmaxIt.second;
+  return std::pair<double, double>(std::min(minEdge, minX),
+                                   std::max(maxEdge, maxX));
+}
+}
+
+namespace Mantid {
+namespace Algorithms {
+
+// Register the algorithm into the AlgorithmFactory
+DECLARE_ALGORITHM(CalculatePolynomialBackground)
+
+//----------------------------------------------------------------------------------------------
+
+/// Algorithms name for identification. @see Algorithm::name
+const std::string CalculatePolynomialBackground::name() const {
+  return "CalculatePolynomialBackground";
+}
+
+/// Algorithm's version for identification. @see Algorithm::version
+int CalculatePolynomialBackground::version() const { return 1; }
+
+/// Algorithm's category for identification. @see Algorithm::category
+const std::string CalculatePolynomialBackground::category() const {
+  return "CorrectionFunctions\\BackgroundCorrections";
+}
+
+/// Algorithm's summary for use in the GUI and help. @see Algorithm::summary
+const std::string CalculatePolynomialBackground::summary() const {
+  return "Fits a polynomial background to a workspace.";
+}
+
+//----------------------------------------------------------------------------------------------
+/** Initialize the algorithm's properties.
+ */
+void CalculatePolynomialBackground::init() {
+  auto increasingAxis = boost::make_shared<API::IncreasingAxisValidator>();
+  auto nonnegativeInt = boost::make_shared<Kernel::BoundedValidator<int>>();
+  nonnegativeInt->setLower(0);
+  auto orderedPairs =
+      boost::make_shared<Kernel::ArrayOrderedPairsValidator<double>>();
+  declareProperty(
+      Kernel::make_unique<API::WorkspaceProperty<API::MatrixWorkspace>>(
+          Prop::INPUT_WS, "", Kernel::Direction::Input, increasingAxis),
+      "An input workspace.");
+  declareProperty(
+      Kernel::make_unique<API::WorkspaceProperty<API::MatrixWorkspace>>(
+          Prop::OUTPUT_WS, "", Kernel::Direction::Output),
+      "A workspace containing the fitted background.");
+  declareProperty(Prop::POLY_DEGREE, 0, nonnegativeInt,
+                  "Degree of the fitted polynomial.");
+  declareProperty(Kernel::make_unique<Kernel::ArrayProperty<double>>(
+                      Prop::XRANGES, std::vector<double>(), orderedPairs),
+                  "A list of fitting ranges given as pairs of X values.");
+}
+
+//----------------------------------------------------------------------------------------------
+/** Execute the algorithm.
+ */
+void CalculatePolynomialBackground::exec() {
```

Comment:
This `exec` method does too much. From a brief glance I can see that there could be separate methods for:
 - Finding the included range
 - Running the fit
 - Pulling the parameters & errors out for the fit workspace
 - Evaluating the background function and filling the Y/E values